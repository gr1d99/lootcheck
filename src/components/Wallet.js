import React from 'react';
import { connect } from 'react-redux';
import { deposit } from "../actions/balance";
import { withdraw } from "../actions/balance";

export const Wallet = (props) => {
	const [newBalance, setBalance] = React.useState(props.balance);

	const handleChange = e => {
		const { value } = e.target;
		setBalance(parseInt(value, 10))
	}

	const handleDeposit = () => {
		props.deposit(newBalance)
	}

	const handleWithdraw = () => {
		props.withdraw(newBalance)
	}

	return (
		<div>
			<h3 className='balance'>
				Wallet Balance: {props.balance}
			</h3>
			<input
				value={newBalance}
				type='number'
				className='input-wallet'
				onChange={handleChange}
			/>
			<button className='btn-deposit' onClick={handleDeposit}>Deposit</button>
			<button className='btn-withdraw' onClick={handleWithdraw}>Withdraw</button>
		</div>
	)
}

export default connect(
	state => { return { balance: state.balance } },
	{ deposit, withdraw }
)(Wallet)
