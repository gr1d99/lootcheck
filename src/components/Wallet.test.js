import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import { Wallet } from './Wallet';

let props;
let wallet;
let balance = 0;
let userBalance = 0;
let mockDeposit;
let mockWithdraw;

const setup = ({ balance = 0 }) => {
	mockDeposit = jest.fn();
	mockWithdraw = jest.fn();
	props = { balance: balance, deposit: mockDeposit, withdraw: mockWithdraw };
	return shallow(<Wallet  {...props} />)
}

describe('<Wallet />', () => {
	beforeEach(() => {
		wallet = setup({ balance });
	})

	it('renders correctly', () => {
		wallet = renderer.create(wallet).toJSON();
		expect(wallet).toMatchSnapshot();
	})

	it('displays the balance from props', () => {
		expect(wallet.find('.balance').text()).toEqual(`Wallet Balance: ${balance}`)
	})

	it('contains an input to deposit or withdraw from balance', () => {
		expect(wallet.find('.input-wallet').exists()).toBe(true)
	})

	describe('When user types into the wallet input', () => {
		userBalance = 20;
		const setBalance = jest.fn();
		const useStateSpy = jest.spyOn(React, 'useState')
		useStateSpy.mockImplementation((init) => [userBalance, setBalance])

		beforeEach(() => {
			wallet
				.find('.input-wallet')
				.simulate('change', { target: { value: userBalance.toString() }})
		})

		it('calls setBalance with `userBalance`', () => {
			expect(setBalance).toHaveBeenCalled()
			expect(setBalance).toHaveBeenCalledWith(userBalance)
			expect(setBalance).toHaveBeenCalledTimes(1)
		})

		describe('and the user wants to make a deposit', () => {
			beforeEach(() => {
				wallet.find('.btn-deposit').simulate('click')
			})

			it('dispatches the `deposit()` it receives from props with the local balance', () => {
				expect(mockDeposit).toHaveBeenCalledTimes(1)
				expect(mockDeposit).toHaveBeenCalledWith(parseInt(userBalance, 10))
			})
		})
	})

	describe('When the user wants to withdraw', () => {
		beforeEach(() => {
			wallet.find('.btn-withdraw').simulate('click')
		})

		it('dispatches the `withdraw()` it receives from props with the local balance', () => {
			expect(mockWithdraw).toHaveBeenCalledWith(parseInt(userBalance, 10))
		})
	})
})
