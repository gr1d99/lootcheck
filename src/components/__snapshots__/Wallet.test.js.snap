// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Wallet /> renders correctly 1`] = `
<div>
  <h3
    className="balance"
  >
    Wallet Balance: 
    0
  </h3>
  <input
    className="input-wallet"
    onChange={[Function]}
    type="number"
    value={20}
  />
  <button
    className="btn-deposit"
    onClick={[Function]}
  >
    Deposit
  </button>
  <button
    className="btn-withdraw"
    onClick={[Function]}
  >
    Withdraw
  </button>
</div>
`;
